example/contrib/socks-client/socks/query.hpp

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright (c) 2018 jackarain (jack dot wgm at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef SOCKS_QUERY_HPP
+#define SOCKS_QUERY_HPP
+
+#include <socks/config.hpp>
+#include <iterator>
+#include <utility>
+
+namespace socks {
+
+/** A ForwardRange representing a reference to a parsed query string.
+*/
+class query
+{
+public:
+    explicit
+    query(string_view s)
+        : s_(s)
+    {
+    }
+
+    class const_iterator
+    {
+    public:
+        using value_type = std::pair<string_view, string_view>;
+        using difference_type = std::ptrdiff_t;
+        using reference = value_type&;
```

Comment:
Actually this is not right, `reference` can't be an actual reference it has to be a value type, e.g.
```
using reference = value_type;
```

And `value_type` can't be a `std::pair` or else the `const_iterator::operator->` won't work (it can't return a pointer). Instead, I suggest you use this declaration for `value_type`:
```
struct value_type
{
  string_view name;
  string_view value;
  value_type const* operator->() const noexcept
  {
    return this;
  }
};
```

Then you can use `using pointer = value_type const;` and things will work right